`shorebird` doesn't notice when engine version mismatches

I had previously run `shorebird build` with the 3.7.7 engine and it was still cached, even though I'd upgrade to 3.7.8 flutter.  When I blew away the .shorebird directory the build fixed itself.

```
eseidel@erics-mbp time_shift % shorebird build
✗ Failed to build: ../flutter/packages/flutter/lib/src/foundation/assertions.dart:675:22: Error: 'NullThrownError' isn't a type.
    if (exception is NullThrownError) { // ignore: deprecated_member_use
                     ^^^^^^^^^^^^^^^
../flutter/packages/flutter/lib/src/gestures/converter.dart:55:14: Error: A non-null value must be returned since the return type 'PointerEvent' doesn't allow null.
 - 'PointerEvent' is from 'package:flutter/src/gestures/events.dart' ('../flutter/packages/flutter/lib/src/gestures/events.dart').
        .map((ui.PointerData datum) {
             ^
Target kernel_snapshot failed: Exception


FAILURE: Build failed with an exception.

* Where:
Script '/Users/eseidel/Documents/GitHub/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 1151

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/eseidel/Documents/GitHub/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 7s
Gradle task bundleRelease failed with exit code 1
 (8.0s)
```

--- a/packages/shorebird_cli/lib/src/auth/auth.dart
+++ b/packages/shorebird_cli/lib/src/auth/auth.dart
@@ -1,16 +1,15 @@
 import 'dart:convert';
 import 'dart:io';
 
-import 'package:cli_util/cli_util.dart';
 import 'package:path/path.dart' as p;
 import 'package:shorebird_cli/src/auth/session.dart';
+import 'package:shorebird_cli/src/config/config.dart';
 
 class Auth {
   Auth() {
     _loadSession();
   }
 
-  static const _applicationName = 'shorebird';
   static const _sessionFileName = 'shorebird-session.json';
 
   void login({required String apiKey}) {
@@ -20,22 +19,11 @@ class Auth {
 
   void logout() => _clearSession();
 
-  late final String? _shorebirdConfigDir = () {
-    try {
-      return applicationConfigHome(_applicationName);
-    } catch (_) {
-      return null;
-    }
-  }();
-
   Session? _session;
 
   Session? get currentSession => _session;
 
   void _loadSession() {
-    final shorebirdConfigDir = _shorebirdConfigDir;
-    if (shorebirdConfigDir == null) return;
-
     final sessionFile = File(p.join(shorebirdConfigDir, _sessionFileName));
 
     if (sessionFile.existsSync()) {
@@ -49,9 +37,6 @@ class Auth {
   }
 
   void _flushSession(Session session) {
-    final shorebirdConfigDir = _shorebirdConfigDir;
-    if (shorebirdConfigDir == null) return;
-
     File(p.join(shorebirdConfigDir, _sessionFileName))
       ..createSync(recursive: true)
       ..writeAsStringSync(json.encode(session.toJson()));
@@ -59,9 +44,6 @@ class Auth {
 
   void _clearSession() {
     _session = null;
-
-    final shorebirdConfigDir = _shorebirdConfigDir;
-    if (shorebirdConfigDir == null) return;
 
     final sessionFile = File(p.join(shorebirdConfigDir, _sessionFileName));
     if (sessionFile.existsSync()) {

--- a/packages/shorebird_cli/lib/src/config/config.dart
+++ b/packages/shorebird_cli/lib/src/config/config.dart
@@ -1,1 +1,2 @@
+export 'shorebird_config_dir.dart';
 export 'shorebird_yaml.dart';

--- a/packages/shorebird_cli/lib/src/config/shorebird_config_dir.dart
+++ b/packages/shorebird_cli/lib/src/config/shorebird_config_dir.dart
@@ -1,0 +1,13 @@
+import 'package:cli_util/cli_util.dart';
+import 'package:meta/meta.dart';
+import 'package:shorebird_cli/src/command_runner.dart';
+
+final String shorebirdConfigDir = () {
+  final configHome = testApplicationConfigHome ?? applicationConfigHome;
+  return configHome(executableName);
+}();
+
+/// Test applicationConfigHome which should
+/// only be used for testing purposes.
+@visibleForTesting
+String Function(String)? testApplicationConfigHome;

--- a/packages/shorebird_cli/lib/src/shorebird_engine_mixin.dart
+++ b/packages/shorebird_cli/lib/src/shorebird_engine_mixin.dart
@@ -5,6 +5,7 @@ import 'dart:isolate';
 import 'package:archive/archive_io.dart';
 import 'package:path/path.dart' as p;
 import 'package:shorebird_cli/src/command.dart';
+import 'package:shorebird_cli/src/config/config.dart';
 import 'package:shorebird_cli/src/flutter_engine_revision.dart';
 import 'package:shorebird_cli/src/shorebird_config_mixin.dart';
 import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
@@ -12,9 +13,9 @@ import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 mixin ShorebirdEngineMixin on ShorebirdConfigMixin {
   String get shorebirdEnginePath {
     return p.join(
-      Directory.current.path,
-      '.shorebird',
-      'engine',
+      shorebirdConfigDir,
+      'engines',
+      requiredFlutterEngineRevision,
     );
   }
 
@@ -22,35 +23,29 @@ mixin ShorebirdEngineMixin on ShorebirdConfigMixin {
   // will not need to cache its own copy of the Shorebird engine.
   Future<void> ensureEngineExists() async {
     final shorebirdEngine = Directory(shorebirdEnginePath);
-    final shorebirdEngineCache = File(
-      p.join(Directory.current.path, '.shorebird', 'cache', 'engine.zip'),
-    );
 
-    if (!shorebirdEngineCache.existsSync()) {
+    if (!shorebirdEngine.existsSync()) {
       final downloadEngineProgress = logger.progress(
         'Downloading shorebird engine',
       );
+      final tempDir = Directory.systemTemp.createTempSync();
+      final engineArchivePath = p.join(tempDir.path, 'engine.zip');
       try {
         final codePushClient = buildCodePushClient(
           apiKey: auth.currentSession!.apiKey,
           hostedUri: hostedUri,
         );
-        await _downloadShorebirdEngine(
-          codePushClient,
-          shorebirdEngineCache.path,
-        );
+        await _downloadShorebirdEngine(codePushClient, engineArchivePath);
         downloadEngineProgress.complete();
       } catch (error) {
         downloadEngineProgress.fail();
         throw Exception('Failed to download shorebird engine: $error');
       }
-    }
 
-    if (!shorebirdEngine.existsSync()) {
       final buildingEngine = logger.progress('Building shorebird engine');
       try {
         await _extractShorebirdEngine(
-          shorebirdEngineCache.path,
+          engineArchivePath,
           shorebirdEngine.path,
           startProcess,
         );

--- a/packages/shorebird_cli/test/src/commands/build_command_test.dart
+++ b/packages/shorebird_cli/test/src/commands/build_command_test.dart
@@ -1,12 +1,14 @@
 import 'dart:io';
 import 'dart:typed_data';
 
+import 'package:archive/archive_io.dart';
 import 'package:args/args.dart';
 import 'package:mason_logger/mason_logger.dart';
 import 'package:mocktail/mocktail.dart';
 import 'package:shorebird_cli/src/auth/auth.dart';
 import 'package:shorebird_cli/src/auth/session.dart';
 import 'package:shorebird_cli/src/commands/build_command.dart';
+import 'package:shorebird_cli/src/config/config.dart';
 import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 import 'package:test/test.dart';
 
@@ -27,6 +29,7 @@ void main() {
     const session = Session(apiKey: 'test-api-key');
 
     late ArgResults argResults;
+    late Directory applicationConfigHome;
     late Auth auth;
     late CodePushClient codePushClient;
     late Logger logger;
@@ -34,6 +37,7 @@ void main() {
     late BuildCommand buildCommand;
 
     setUp(() {
+      applicationConfigHome = Directory.systemTemp.createTempSync();
       argResults = _MockArgResults();
       auth = _MockAuth();
       codePushClient = _MockCodePushClient();
@@ -49,6 +53,7 @@ void main() {
           return processResult;
         },
       )..testArgResults = argResults;
+      testApplicationConfigHome = (_) => applicationConfigHome.path;
 
       when(() => argResults.rest).thenReturn([]);
       when(
@@ -84,9 +89,12 @@ void main() {
       when(() => processResult.exitCode).thenReturn(1);
       when(() => processResult.stderr).thenReturn('oops');
       final tempDir = Directory.systemTemp.createTempSync();
-      Directory('${tempDir.path}/.shorebird/engine')
-          .createSync(recursive: true);
       when(() => auth.currentSession).thenReturn(session);
+      when(
+        () => codePushClient.downloadEngine(revision: any(named: 'revision')),
+      ).thenAnswer(
+        (_) async => Uint8List.fromList(ZipEncoder().encode(Archive())!),
+      );
 
       final result = await IOOverrides.runZoned(
         () async => buildCommand.run(),
@@ -99,12 +107,11 @@ void main() {
     test('exits with code 0 when building succeeds', () async {
       when(() => processResult.exitCode).thenReturn(ExitCode.success.code);
       final tempDir = Directory.systemTemp.createTempSync();
-      Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
       when(
         () => codePushClient.downloadEngine(revision: any(named: 'revision')),
-      ).thenAnswer((_) async => Uint8List.fromList([]));
+      ).thenAnswer(
+        (_) async => Uint8List.fromList(ZipEncoder().encode(Archive())!),
+      );
       when(() => auth.currentSession).thenReturn(session);
 
       final result = await IOOverrides.runZoned(

--- a/packages/shorebird_cli/test/src/commands/publish_command_test.dart
+++ b/packages/shorebird_cli/test/src/commands/publish_command_test.dart
@@ -1,6 +1,7 @@
 import 'dart:io';
 import 'dart:typed_data';
 
+import 'package:archive/archive.dart';
 import 'package:args/args.dart';
 import 'package:mason_logger/mason_logger.dart';
 import 'package:mocktail/mocktail.dart';
@@ -8,6 +9,7 @@ import 'package:path/path.dart' as p;
 import 'package:shorebird_cli/src/auth/auth.dart';
 import 'package:shorebird_cli/src/auth/session.dart';
 import 'package:shorebird_cli/src/commands/publish_command.dart';
+import 'package:shorebird_cli/src/config/config.dart';
 import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 import 'package:test/test.dart';
 
@@ -58,6 +60,7 @@ flutter:
     - shorebird.yaml''';
 
     late ArgResults argResults;
+    late Directory applicationConfigHome;
     late Auth auth;
     late Progress progress;
     late Logger logger;
@@ -79,6 +82,7 @@ flutter:
 
     setUp(() {
       argResults = _MockArgResults();
+      applicationConfigHome = Directory.systemTemp.createTempSync();
       auth = _MockAuth();
       progress = _MockProgress();
       logger = _MockLogger();
@@ -95,6 +99,7 @@ flutter:
         },
         logger: logger,
       )..testArgResults = argResults;
+      testApplicationConfigHome = (_) => applicationConfigHome.path;
 
       when(() => argResults.rest).thenReturn([]);
       when(() => auth.currentSession).thenReturn(session);
@@ -190,13 +195,13 @@ flutter:
       when(() => processResult.stderr).thenReturn('oops');
       when(() => auth.currentSession).thenReturn(session);
 
+      when(
+        () => codePushClient.downloadEngine(revision: any(named: 'revision')),
+      ).thenAnswer(
+        (_) async => Uint8List.fromList(ZipEncoder().encode(Archive())!),
+      );
+
       final tempDir = setUpTempDir();
-      Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
-      ).createSync(recursive: true);
       final exitCode = await IOOverrides.runZoned(
         () async => command.run(),
         getCurrentDirectory: () => tempDir,
@@ -209,10 +214,7 @@ flutter:
         () async {
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final exitCode = await IOOverrides.runZoned(
         command.run,
@@ -226,15 +228,10 @@ flutter:
 
     test('throws error when fetching apps fails.', () async {
       const error = 'something went wrong';
-      when(
-        () => codePushClient.getApps(),
-      ).thenThrow(error);
+      when(() => codePushClient.getApps()).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -268,10 +265,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -302,10 +296,7 @@ flutter:
       when(() => codePushClient.getApps()).thenAnswer((_) async => []);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -335,10 +326,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -375,10 +363,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -411,10 +396,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -453,10 +435,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -486,10 +465,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -525,10 +501,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -564,10 +537,7 @@ flutter:
       ).thenThrow(error);
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -593,10 +563,7 @@ flutter:
     test('succeeds when publish is successful', () async {
       final tempDir = setUpTempDir();
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,
@@ -631,10 +598,7 @@ app_id: $appId
 base_url: $baseUrl''',
       );
       Directory(
-        '${tempDir.path}/.shorebird/engine',
-      ).createSync(recursive: true);
-      Directory(
-        '${tempDir.path}/.shorebird/cache',
+        p.join(command.shorebirdEnginePath, 'engine'),
       ).createSync(recursive: true);
       final artifactPath = p.join(
         tempDir.path,

--- a/packages/shorebird_cli/test/src/commands/run_command_test.dart
+++ b/packages/shorebird_cli/test/src/commands/run_command_test.dart
@@ -11,6 +11,7 @@ import 'package:path/path.dart' as p;
 import 'package:shorebird_cli/src/auth/auth.dart';
 import 'package:shorebird_cli/src/auth/session.dart';
 import 'package:shorebird_cli/src/commands/run_command.dart';
+import 'package:shorebird_cli/src/config/config.dart';
 import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 import 'package:test/test.dart';
 
@@ -31,6 +32,7 @@ void main() {
     const session = Session(apiKey: 'test-api-key');
 
     late ArgResults argResults;
+    late Directory applicationConfigHome;
     late Auth auth;
     late Logger logger;
     late Process process;
@@ -39,6 +41,7 @@ void main() {
 
     setUp(() {
       argResults = _MockArgResults();
+      applicationConfigHome = Directory.systemTemp.createTempSync();
       auth = _MockAuth();
       logger = _MockLogger();
       process = _MockProcess();
@@ -53,6 +56,8 @@ void main() {
           return process;
         },
       )..testArgResults = argResults;
+
+      testApplicationConfigHome = (_) => applicationConfigHome.path;
 
       when(() => argResults.rest).thenReturn([]);
       when(() => logger.progress(any())).thenReturn(_MockProgress());
@@ -92,9 +97,6 @@ void main() {
 
     test('exits with code 70 when building the engine fails', () async {
       final tempDir = Directory.systemTemp.createTempSync();
-      Directory(
-        p.join(tempDir.path, '.shorebird', 'cache'),
-      ).createSync(recursive: true);
 
       when(() => auth.currentSession).thenReturn(session);
       when(
@@ -119,12 +121,9 @@ void main() {
 
     test('exits with code when running the app fails', () async {
       final tempDir = Directory.systemTemp.createTempSync();
-      final engineCacheDir = Directory(
-        p.join(tempDir.path, '.shorebird', 'cache'),
-      )..createSync(recursive: true);
 
       ZipFileEncoder()
-        ..create(p.join(engineCacheDir.path, 'engine.zip'))
+        ..create(p.join(runCommand.shorebirdEnginePath, 'engine.zip'))
         ..close();
 
       when(() => auth.currentSession).thenReturn(session);
@@ -157,14 +156,10 @@ void main() {
 
     test('exits with code 0 when running the app succeeds', () async {
       final tempDir = Directory.systemTemp.createTempSync();
-      Directory(p.join(tempDir.path, '.shorebird', 'cache'))
-          .createSync(recursive: true);
-      Directory(p.join(tempDir.path, '.shorebird', 'engine'))
-          .createSync(recursive: true);
+      Directory(
+        p.join(runCommand.shorebirdEnginePath, 'engine'),
+      ).createSync(recursive: true);
       when(() => auth.currentSession).thenReturn(session);
-      when(
-        () => codePushClient.downloadEngine(revision: any(named: 'revision')),
-      ).thenAnswer((_) async => Uint8List(0));
 
       final progress = _MockProgress();
       when(() => logger.progress(any())).thenReturn(progress);

